serve/mlc_serve/model/paged_cache_model.py

```diff
@@ -301,22 +309,38 @@ def _is_safe_to_sample(prob_like):
         logits = _apply_top_p_top_k(logits_random, top_ps, top_ks)
 
     probs = torch.softmax(logits_random, dim=-1)
+    logprobs = torch.log(torch.softmax(logits_greedy, dim=-1))
+    top_random_logprob, top_random = torch.topk(logprobs, k=5, dim=-1, largest=True, sorted=True)
+    top_random_logprob = top_random_logprob.cpu().numpy()
+    top_random = top_random.cpu().numpy()
 
     if check_safety and not _is_safe_to_sample(probs):
         return None
 
     res_random = torch.multinomial(probs, 1, True).cpu().numpy()[:, 0]
+    res_random_logprobs = torch.gather(logprobs, dim=-1, index=torch.tensor(res_random, dtype=torch.int64, device=logits.device)).cpu().numpy()
 
     if logits_random.shape[0] == num_seq:
-        return res_random
+        return res_random, (res_random_logprobs, (top_random, top_random_logprob))
 
     res = np.empty((num_seq,), dtype=np.int32)
+    res_logprobs = np.empty((num_seq,), dtype=np.float32)
+    top = np.empty((num_seq, 5), dtype=np.int32)
+    top_logprobs = np.empty((num_seq, 5), dtype=np.float32)
+
     res[mask_random] = res_random
+    res_logprobs[mask_random] = res_random_logprobs
+    top[mask_random] = top_random
+    top_logprobs[mask_random] = top_random_logprob
+
 
     if logits_greedy.shape[0] > 0:
         res[mask_greedy] = res_greedy
+        res_logprobs[mask_greedy] = res_greedy_logprob
+        top[mask_greedy] = top_greedy
+        top_logprobs[mask_greedy] = top_greedy_logprob
 
-    return res
+    return res, ((res, res_logprobs), (top, top_logprobs))
```

Comment:
If logprob is not frequently asked, I don't want to pay the cost of logprob calculation when it is not needed. Can we decouple the logprob calculation from this change and do it only when necessary?

Reply:
I think it's possible to decouple logprob calculation from original sampling, only do it when necessary. During batching I'm thinking of separating queries with logprobs and without. On API side I would like to keep it this way and return a `None` when no logprob is requred.

Reply:
One thought. Maybe we can introduce a debug mode of engine execution that enables logprobs and outputs rich debug info. For prod mode, it shouldn't take any effect.